osu.Game/Screens/Select/Carousel/TopLocalRank.cs

```diff
@@ -30,72 +30,45 @@ public class TopLocalRank : UpdateableRank
         [Resolved]
         private IAPIProvider api { get; set; }
 
+        private IDisposable scoreSubscription;
+
+        private bool rankUpdatePending;
+
         public TopLocalRank(BeatmapInfo beatmapInfo)
             : base(null)
         {
             this.beatmapInfo = beatmapInfo;
         }
 
-        [BackgroundDependencyLoader]
-        private void load()
-        {
-            ruleset.ValueChanged += _ => fetchAndLoadTopScore();
-
-            fetchAndLoadTopScore();
-        }
-
         protected override void LoadComplete()
         {
             base.LoadComplete();
 
-            scoreSubscription = realmFactory.Context.All<ScoreInfo>()
-                                            .Filter($"{nameof(ScoreInfo.BeatmapInfo)}.{nameof(BeatmapInfo.ID)} = $0", beatmapInfo.ID)
-                                            .QueryAsyncWithNotifications((_, changes, ___) =>
-                                            {
-                                                if (changes == null)
-                                                    return;
-
-                                                fetchTopScoreRank();
-                                            });
-        }
-
-        private IDisposable scoreSubscription;
-
-        private ScheduledDelegate scheduledRankUpdate;
-
-        private void fetchAndLoadTopScore()
-        {
-            // TODO: this lookup likely isn't required, we can use the results of the subscription directly.
-            var rank = fetchTopScoreRank();
-
-            scheduledRankUpdate = Scheduler.Add(() =>
+            ruleset.BindValueChanged(_ =>
             {
-                Rank = rank;
-
+                rankUpdatePending = true;
                 // Required since presence is changed via IsPresent override
                 Invalidate(Invalidation.Presence);
-            });
+
+                scoreSubscription?.Dispose();
+                scoreSubscription = realmFactory.Context.All<ScoreInfo>()
+                                                .Filter($"{nameof(ScoreInfo.User)}.{nameof(RealmUser.OnlineID)} == $0"
+                                                        + $" && {nameof(ScoreInfo.BeatmapInfo)}.{nameof(BeatmapInfo.ID)} == $1"
+                                                        + $" && {nameof(ScoreInfo.Ruleset)}.{nameof(RulesetInfo.ShortName)} == $2"
+                                                        + $" && {nameof(ScoreInfo.DeletePending)} == false", api.LocalUser.Value.Id, beatmapInfo.ID, ruleset.Value.ShortName)
+                                                .OrderByDescending(s => s.TotalScore)
+                                                .QueryAsyncWithNotifications((items, changes, ___) =>
+                                                {
+                                                    if (changes == null)
+                                                        rankUpdatePending = false;
```

Comment:
Is the `changes` null check there just for the initial update? If so, I think that hoisting the presence invalidation inside the callback here will render this conditional and the `rankUpdatePending` field itself no longer needed.

In fact the comment on the `IsPresent` override is mentioning the drawable's scheduler which is no longer used in this flow, I think? Realm uses the sync context installed in framework, which in turn uses the game-global scheduler rather than the drawable's (although that may be changing at some point? but I'm not sure that's a justification to keep this code in its current shape).

Reply:
I've removed all of this, and also added test coverage.